### ISSUE TITLE
Variable scoping

### ISSUE THREAD

**Issue by derekparker** [derekparker/delve#106]
Given the following source:

```go
func fn() {
        for i := 0; i < 10: i++ {
                fmt.Println(i)
        }

        for i := 0; i < 10; i++ {
            fmt.Println(i)
        }
}
```

If you break in the second loop, the result of `print i` will be 10 instead of the value of the next `i` that shadows it.

**Comment by derekparker** [comment]
@epipho would you have any interest in taking a look at this?

**Comment by epipho** [comment]
Yeah I will take a look at this.  I think I know what is going on.

**Comment by derekparker** [comment]
@epipho awesome, thank you.

**Comment by epipho** [comment]
So this is what I thought and is unfortunately a related problem to #83 

There are actually two variables with the same name and we print the first one we come across.  You can actually see this if you use info locals, it will show both.

```
   6:   for foobarbaz := 0; foobarbaz < 10; foobarbaz++ {
   7:           fmt.Println(foobarbaz)
   8:   }
   9:
   10:  for foobarbaz := 0; foobarbaz < 10; foobarbaz++ {
=> 11:          fmt.Println(foobarbaz)
   12:  }
   13: }

(dlv) info locals
foobarbaz = 0
foobarbaz = 10
(dlv) print foobarbaz
10
```

As in #83 there is no way to disambiguate them as there is no scoping or .debug_loc data

```
 <2><40>: Abbrev Number: 4 (DW_TAG_variable)
    <41>   DW_AT_name        : foobarbaz
    <4b>   DW_AT_location    : 5 byte block: 9c 11 98 7e 22     (DW_OP_call_frame_cfa; DW_OP_consts: -232; DW_OP_plus)
    <51>   DW_AT_type        : <0x29a43>
 <2><59>: Abbrev Number: 4 (DW_TAG_variable)
    <5a>   DW_AT_name        : foobarbaz
    <64>   DW_AT_location    : 5 byte block: 9c 11 a0 7e 22     (DW_OP_call_frame_cfa; DW_OP_consts: -224; DW_OP_plus)
    <6a>   DW_AT_type        : <0x29a43>
```

For print statements we could print out all the variables with the name, but the order is not going to be guaranteed.

Thoughts? I think we really need to solve this issue for a 1.0, but am not sure where to start other than putting together a proposal for go1.6 as 1.5 is about to lock.

**Comment by derekparker** [comment]
@epipho thanks for looking into this. You're right, this seems to be a Go linker shortcoming. 

Let's put together a proposal and get it submitted as soon as possible.

**Comment by ironcladlou** [comment]
Not that I necessarily think this was controversial, but just wanted to chime in and say I too think this is nasty enough to block a 1.0. Unfortunate!

**Comment by james-lawrence** [comment]
@epipho are you sure?
DW_AT_location might be usable. not sure what each value in the output you showed represents.
but the location block for one is higher that the other, might be able to resolve it based on that. (not in anyway familiar with the dwarf spec. just eyeballing the output you displayed.)

couple spots might be worth looking into what they mean:
<2><40>: Abbrev Number vs <2><59>: Abbrev Number
<4b> DW_AT_location vs <64> DW_AT_location
5 byte block: 9c 11 98 7e 22 vs 5 byte block: 9c 11 a0 7e 22
